hypha/apply/todo/views.py

```diff
@@ -0,0 +1,138 @@
+from django.contrib.contenttypes.models import ContentType
+from django.db.models import Count
+
+from .models import Task
+from .options import get_task_template
+from .services import validate_user_groups_uniqueness, validate_user_uniquness
+
+
+def add_task_to_user(code, user, related_obj):
+    """
+    Add task for a user
+    input:
+        code: TASKS_CODE_CHOICES.keys()
+        user: User object
+        related_obj: Object - Submission, Project, Invoice, Report
+    output: task - Task object / None in case of no creation
+    """
+    user_uniqueness = validate_user_uniquness(
+        code=code, user=user, related_obj=related_obj
+    )
+    if user_uniqueness:
+        task = Task.objects.create(code=code, user=user, related_object=related_obj)
+        return task
+    return None
+
+
+def add_task_to_user_group(code, user_group, related_obj):
+    """
+    Add task for user_groups
+    input:
+        code: TASKS_CODE_CHOICES.keys()
+        user_group: Queryset - Group objects
+        related_obj: Object - Submission, Project, Invoice, Report
+    output: task - Task object / None in case of no creation
+    """
+    user_groups_uniqueness = validate_user_groups_uniqueness(
+        code=code, user_groups=user_group, related_obj=related_obj
+    )
+    if user_groups_uniqueness:
+        task = Task.objects.create(code=code, related_object=related_obj)
+        groups = [Group.objects.filter(id=group.id).first() for group in user_group]
+        task.user_group.add(*groups)
+        return task
+    return None
+
+
+def remove_tasks_for_user(code, user, related_obj):
+    """
+    Remove task for a user
+    input:
+        code: TASKS_CODE_CHOICES.keys()
+        user: User object
+        related_obj: Object - Submission, Project, Invoice, Report
+    output: None
+    """
+    task = Task.objects.filter(
+        code=code,
+        user=user,
+        related_content_type=ContentType.objects.get_for_model(related_obj).id,
+        related_object_id=related_obj.id,
+    ).first()
+    if task:
+        task.delete()
+    return None
+
+
+def remove_tasks_for_user_group(code, user_group, related_obj):
+    """
+    Remove task for user_groups
+    input:
+        code: TASKS_CODE_CHOICES.keys()
+        user_group: Queryset - Group objects
+        related_obj: Object - Submission, Project, Invoice, Report
+    output: None
+    """
+    matching_tasks = Task.objects.filter(
+        code=code,
+        related_content_type=ContentType.objects.get_for_model(related_obj).id,
+        related_object_id=related_obj.id,
+    )
+    user_group_matching_tasks = matching_tasks.annotate(
+        group_count=Count("user_group")
+    ).filter(group_count=len(user_group.all()))
+    for group in user_group.all():
+        user_group_matching_tasks = user_group_matching_tasks.filter(
+            user_group__id=group.id
+        )
+    if user_group_matching_tasks.exists():
+        user_group_matching_tasks.delete()
```

Comment:
This looks is over-writing the `user_group_matching_tasks` in every iteration. You are then removing the last one. Something is not adding up. 

Would be great if you can write a test to check it's going what it should do.

Reply:
It is just finding the tasks with exact user groups(many to many relations) and in the end, deleting those tasks.

I'll add tests for all of these APIs later.  As this PR already has too many changes, I'll add a separate PR for that.

Reply:
It's okay to have the PR as part of this PR itself. Also can you clarify bit more what you mean by `in the end`. 

You are looping over something and just picking the last value for that iteration. I don't see it being predictable. 

image user_group.all() returns `group 1` and `group 2`, you are going be acting on the `group 2` and not do anything for `group 1`. If that is case, you can just `user_group.last()` instead of `user_group.all()`

Reply:
Okay, Let me explain the loop logic here. As I mentioned in my last comment it aims to find the tasks with 'exact' user_groups.

Ex: Let's say there are two groups `group1` and `group2` in user_group.
**In line 82:** annotate and filter is making sure that `user_group_matching_tasks` contains all the tasks with two groups(len of 'user_group'), but we are not sure these groups are exclusively `group1` and `group2`.
So for that, we iterate through 'user_group'. In the first iteration, the code filters `user_group_matching_tasks` to include only those with `group1` and updates the collection accordingly. In the second iteration, it ensures that the 'updated' `user_group_matching_tasks` collection also contains `group2`. 

After all iterations, we can be sure that all tasks in `user_group_matching_tasks` have exactly two groups, specifically `group1` and `group2`.